salvo/ENVOY_DEVELOP_WORKFLOW.md

```diff
@@ -0,0 +1,135 @@
+# Measure Envoy's Performance Change with an A/B Testing
+
+## Objective 
+
+This documentation will describe the user journey and detailed workflow to use Salvo to measure Envoy's performance change with an A/B testing.
+
+## User Journey
+
+You are implementing a new feature in Envoy and reach a point where you need to decide between two possible implementation approaches. You wonder what are the performance implications of each choice and decide to run the benchmarking framework to find out. You run two executions in A/B testing mode, both comparing an official Envoy version to an Envoy version with your change using one of the approaches. Once the tests complete, you examine the results and choose an implementation approach with the better performance characteristics.
+
+## Workflow
+
+This section will describe the whole workflow to use Salvo to Measure Performance Change with an A/B Testing. And we will use [Binary Benchmark](./README.md#binary-benchmark) as an example.
+
+1. The requirements to build Envoy and Nighthawk in local machine
+
+    In oder to build Envoy and Nighthawk in local machine, follow [this documentation](https://www.envoyproxy.io/docs/envoy/latest/start/building.html#linux-mac-target-requirements) to ensure local machine meets the requirements. 
+
+2. Download Salvo
+
+    The next step is to download Salvo project from github, you can use `git clone` command:
+
+    ```bash
+    git clone https://github.com/envoyproxy/envoy-perf.git
+    cd envoy-perf/salvo
+    ```
+
+3. Install dependencies 
+
+    Before using Salvo, you need to install its dependencies. The [install_deps.sh](./install_deps.sh) script can be used to install any dependencies required by Salvo. 
+
+    ```bash
+    ./install_deps.sh
+    ```
+
+4. Build Salvo
+
+    The next step is to build Salvo, Salvo can be built by [Bazel](https://docs.bazel.build/versions/main/install-ubuntu.html):
+
+    ```bash
+    bazel build //...
+    ```
+
+5. Binary Benchmark Configuration
+    
+    For an example, you committed two Envoy code changes to your forked repo "https://github.com/gyohuangxin/envoy" and the commit hashes were "c2e1c0d0da61f3fec25f48f8bd63e4a76db8a2a9" and "e95717bcd0672910826de4a9bb2f402f6e057b2c". Then you want to measure their performance change to Envoy v1.21.0 as the baseline and choose the commit which has the better performance. 
+    
+    You can compose the configuration of Salvo as below:
+
+    ```yaml
+    remote: false
+    binaryBenchmark: true
+    environment:
+      outputDir: /home/ubuntu/nighthawk_output
+      testVersion: IPV_V4ONLY
+    images:
+      nighthawkBenchmarkImage: envoyproxy/nighthawk-benchmark-dev:latest
+    source:
+    - identity: SRCID_ENVOY
+      commit_hash: v1.21.0
+      additional_hashes: [c2e1c0d0da61f3fec25f48f8bd63e4a76db8a2a9, e95717bcd0672910826de4a9bb2f402f6e057b2c]
+      source_url: https://github.com/gyohuangxin/envoy.git
```

Comment:
I think I've explained that in the doc above: 
```
For an example, you committed two Envoy code changes to your forked repo "https://github.com/gyohuangxin/envoy" and the commit hashes were "c2e1c0d0da61f3fec25f48f8bd63e4a76db8a2a9" and "e95717bcd0672910826de4a9bb2f402f6e057b2c". Then you want to measure their performance change to Envoy v1.21.0 as the baseline and choose the commit which has the better performance.
```
The doc is for developers who has two commits in their own repo already, I think, it's not necessary to write the process of push commits.

Reply:
Thank you for pointing that out. I actually missed that part / didn't create the connection, probably because the text is written very passively.

Would it help if we rephrase it into more assertive / active instructions? Maybe even split into its own step that would clearly say what the reader should do. E.g.:

5. Set up commits for evaluation

Create a local fork of the Envoy repository and commit changes whose performance you want to evaluate. For example ... this is where we can continue with the rest of the example.

This is also where we can indicate that the example in this document is from the main Envoy repository to make it reusable.

7. Binary Benchmark Configuration

The rest of the configuration step.

Reply:
It's a really great idea. Maybe I have some misunderstanding about the user story. In my previous opinion, the developers has commits already in their forked repo. But in fact, they probably has not commits and just write their codes locally. Your comments let me realize the problem.

However, if I follow this flow, we still cannot indicate a real commit as an example. Can we still use the virtual commit as the example which just clarifies the workflow? And then I can create an other PR to introduce the real commit from Envoy repository after I try commits you mention above. Because it may take a period of time to try the real commits, which will make this PR pending for a while.

Reply:
Taking this step by step makes sense, we don't want to keep this PR waiting for a long time.

We can go with fake commits in this PR, but let's make that clear in the text. You can then send a follow-up PR that will convert the example to something that can be executed as written.

Reply:
@mum4k Thanks, I'll make it clear.